test/hummingbot/connector/exchange/ascend_ex/test_ascend_ex_api_order_book_data_source.py

```diff
@@ -0,0 +1,582 @@
+        await self.data_source._shared_client.close()
+        return sent_messages
+
+    def test_listen_for_subscriptions_registers_to_orders_trades_and_instruments(self):
+        sent_messages = self.ev_loop.run_until_complete(self.do_listen_for_subscriptions_registers_to_orders_trades_and_instruments())
+
+        self.assertEqual(2, len(sent_messages))
+        expected_trades_subscription = {'op': 'sub', 'ch': 'trades:BTC/USDT'}
+        expected_depth_subscription = {'op': 'sub', 'ch': 'depth:BTC/USDT'}
+        self.assertEqual(expected_trades_subscription, sent_messages[0])
+        self.assertEqual(expected_depth_subscription, sent_messages[1])
+
+    @patch('aiohttp.ClientSession.ws_connect', new_callable=AsyncMock)
+    async def do_listen_for_order_book_diff_raises_cancel_exceptions(self, ws_connect_mock):
+        self.data_source._shared_client = aiohttp.ClientSession()
+        ws_connect_mock.return_value = self.mocking_assistant.create_websocket_mock()
+
+        queue = asyncio.Queue()
+        self.listening_task = self.ev_loop.create_task(self.data_source.listen_for_order_book_diffs(ev_loop=self.ev_loop, output=queue))
+
+        with self.assertRaises(asyncio.CancelledError):
+            self.listening_task.cancel()
+            await self.listening_task
+
+        await self.data_source._shared_client.close()
+
+    def test_listen_for_order_book_diff_raises_cancel_exceptions(self):
+        self.ev_loop.run_until_complete(self.do_listen_for_order_book_diff_raises_cancel_exceptions())
+
+    @patch('aiohttp.ClientSession.ws_connect', new_callable=AsyncMock)
+    async def do_listen_for_order_book_diff_raises_cancel_exception_when_canceled_during_ws_connection(self, ws_connect_mock):
+        self.data_source._shared_client = aiohttp.ClientSession()
+        ws_connect_mock.side_effect = asyncio.CancelledError()
+
+        queue = asyncio.Queue()
+        self.listening_task = self.ev_loop.create_task(self.data_source.listen_for_order_book_diffs(ev_loop=self.ev_loop, output=queue))
+
+        with self.assertRaises(asyncio.CancelledError):
+            await self.listening_task
+
+        await self.data_source._shared_client.close()
+
+    def test_listen_for_order_book_diff_raises_cancel_exception_when_canceled_during_ws_connection(self):
+        self.ev_loop.run_until_complete(self.do_listen_for_order_book_diff_raises_cancel_exception_when_canceled_during_ws_connection())
+
+    @patch('aiohttp.ClientSession.ws_connect', new_callable=AsyncMock)
+    async def do_listen_for_order_book_diff_ws_connection_exception_details_are_logged(self, ws_connect_mock):
+        self.data_source._shared_client = aiohttp.ClientSession()
+        ws_connect_mock.side_effect = Exception()
+
+        queue = asyncio.Queue()
+        self.listening_task = self.ev_loop.create_task(self.data_source.listen_for_order_book_diffs(ev_loop=self.ev_loop, output=queue))
+        await asyncio.sleep(0.5)
```

Comment:
Here also instead of sleeping we should wait for a particular message to arrive. That can be achieved by making the side effect raise an exception only the first time, but the second time returning a valid response.